central/globaldb/v2backuprestore/restore/postgres.go

```diff
@@ -0,0 +1,222 @@
+	if err != nil {
+		log.Errorf("Unable to drop database %s", databaseName)
+		return err
+	}
+
+	return nil
+}
+
+// createDB - creates a temporary database from template for restoring
+func createDB(sourceMap map[string]string, config *pgxpool.Config) error {
+	// Set the options for pg_dump from the connection config
+	options := []string{
+		"-T",
+		createTemplate,
+		restoreDB,
+	}
+
+	// Get the common DB connection info
+	options = append(options, common.GetConnectionOptions(config)...)
+
+	cmd := exec.Command("createdb", options...)
+
+	common.SetPostgresCmdEnv(cmd, sourceMap, config)
+
+	return common.ExecutePostgresCmd(cmd)
+}
+
+// LoadRestoreStream a Postgres database from a dump
+func LoadRestoreStream(fileReader io.Reader) error {
+	log.Info("Starting Postgres Restore")
+
+	sourceMap, config, err := globaldb.GetPostgresConfig()
+	if err != nil {
+		return errors.Wrap(err, "Could not parse postgres config")
+	}
+
+	// Now recreate the DB
+	err = createDB(sourceMap, config)
+	if err != nil {
+		return errors.Wrap(err, "Could not create restore database")
+	}
+
+	// Execute the restore on the temporary restore database
+	err = runRestoreStream(fileReader, sourceMap, config)
+	if err != nil {
+		return errors.Wrap(err, "Could not load the Postgres backup")
+	}
+
+	log.Info("Postgres Restore Complete")
+	return nil
+}
+
+func runRestoreStream(fileReader io.Reader, sourceMap map[string]string, config *pgxpool.Config) error {
+	// Set the options for pg_dump from the connection config
+	options := []string{
+		"-d",
+		restoreDB,
+		"--no-owner",
+		"--clean",
+		"--if-exists",
+		"--exit-on-error",
+		"-Fc",
+		"-vvv",
+		"--single-transaction",
+	}
+
+	// Get the common DB connection info
+	options = append(options, common.GetConnectionOptions(config)...)
+
+	cmd := exec.Command("pg_restore", options...)
+
+	// Set stdin to be the incoming reader
+	cmd.Stdin = fileReader
+
+	common.SetPostgresCmdEnv(cmd, sourceMap, config)
+	err := common.ExecutePostgresCmd(cmd)
+
+	if err != nil {
+		// Clean up the restore DB since the restore failed
+		_ = dropDB(sourceMap, config, restoreDB)
+		return errors.Wrap(err, "Unable to restore the postgres dump")
+	}
+
+	return nil
+}
+
+func renameRestoreDB(connectPool *pgxpool.Pool, updatedDB, primaryDB string) error {
+	log.Info("Flipping the restored DB to the active DB")
+	sqlStmt := fmt.Sprintf("ALTER DATABASE %s RENAME TO %s", updatedDB, primaryDB)
+
+	_, err := connectPool.Exec(context.TODO(), sqlStmt)
+
+	return err
+}
+
+// SwitchToRestoredDB - switches the restore DB to be the active DB
+func SwitchToRestoredDB(sourceMap map[string]string, config *pgxpool.Config) error {
+	log.Info("Switching to restored database")
+
+	// Connect to different database for admin functions
+	connectPool := adminPool(config)
+
+	// Restore succeeded to the separate DB, so we need to drop the original in order to rename
+	// the new one.
+	err := dropDB(sourceMap, config, config.ConnConfig.Database)
```

Comment:
We should keep original db in case we need to switch back.

Reply:
We can do that as part of restore and always keep 2 OR we can deal with that as part of https://issues.redhat.com/browse/ROX-9884 when we have a better plan on how we would want to do rollbacks.  I think I would prefer to do it under that ticket than this one.